crates/engine/tree/src/tree/mod.rs

```diff
@@ -875,6 +913,49 @@ impl PersistenceState {
     }
 }
 
+/// Represents the tree state kept in memory.
+trait InMemoryState {
+    /// Returns the state for a given block hash.
+    fn in_memory_state_by_hash(&self, hash: B256) -> Option<Arc<State>>;
+    /// Returns the state for a given block number.
+    fn in_memory_state_by_number(&self, number: u64) -> Option<Arc<State>>;
+    /// Returns the current chain head.
+    fn in_memory_current_head(&self) -> Option<(BlockNumber, B256)>;
+    /// Returns the pending block hash.
+    fn in_memory_pending_block_hash(&self) -> Option<B256>;
+    /// Returns the pending state corresponding to the current head plus one, if
+    /// we have the payload.
```

Comment:
what does "if we have the payload" mean here?

Reply:
comes from the issue description https://github.com/paradigmxyz/reth/issues/8747 As I understand it, it might mean that we are the proposers for this slot and have received the payload attributes as part of the fcu call, is that correct @mattsse ?